### ISSUE TITLE
support reading data from places which aren't files

### ISSUE THREAD

**Issue by NicklasTegner** [synthizer/synthizer#11]
Hi.

I would suggest prioritizing a "memory" protocol for synthizer buffers.
My reasoning is, that with a "file" protocol and a "memory" protocol, almost every generic use case would be covered, since you can **almost** convert every other use case (that I can think of) to either a file, or in-memory data.

thoughts?

**Comment by chrisnorman7** [comment]
I agree. That would wrap up grabbing sound data - from the internet for example - quite nicely. Unless I'm missing something, which is entirely possible.

**Comment by camlorn** [comment]
This is prioritized for after the planned HRTF improvement and basic reverb.  What you'll be getting is the ability to wire in your own stream protocols directly.  In C this'll be callbacks:

```
int readFromStream(void *stream) {
    // ...
}

bool seekInStream(void* stream, int pos) {
    //...
}
```

Etc.  And in Python, you'll do it by inheriting from a class.

Then things like reading from the internet, etc, are you just do them and don't round-trip through a temporary buffer that you construct only for the sake of giving it to a Synthizer decode function.

There will probably then be a simplified helper for buffers themselves that doesn't require going through the trouble of registering a custom protocol when you already have a chunk of memory you want to feed in, but the above lets you do lots of stuff and avoids having a big wav file fully loaded into memory twice over and also works with streaming, etc.

**Comment by ndarilek** [comment]
Not sure if this also includes the possibility of handing Synthizer an array of u8s and a content type, and having it create a buffer from that? If not then I'll create a separate issue. I'd like to create an asset loader for [Bevy](https://bevyengine.org), and it expects to hand something an array of bytes and get back an arbitrarily-typed asset.

I suspect this is under the same umbrella, but I don't really need seeking and a callback interface would be overkill. Perhaps that's where the helpers come in, but in the off-chance that I'm wrong, I wanted to make sure this use case was covered.

Thanks.

**Comment by camlorn** [comment]
@ndarilek 
The short answer is yes.  The long answer is also yes, pretty much.  Though the helper is specifically for doing what you want to do, being as the underlying streaming thing is more complicated.  No need to make you implement 5 methods for loading a buffer.

Do you need the ability to manually specify the content type?  At the moment Synthizer just autodetects it and I wasn't planning to really make that change.

**Comment by ndarilek** [comment]
Cool deal. Content-type autodetection is fine. I just need to hand 
Synthizer some bytes and have it hand me back a buffer it can play. Thanks!

**Comment by camlorn** [comment]
You're welcome, also the timeline at the moment is probably ~3 weeks.  This one is after #32 which became super super high priority after Linux support because reasons, and is a pretty big refactor.  But this ticket is maybe the number 1 requested thing, so after performance stops sucking, etc.

**Comment by ahicks92** [comment]
It obviously took longer than 3 weeks, but Synthizer 0.9 will have `syz_createBufferFromEncodedData`, which is implemented now, as well as `syz_createBufferFromFloat`.  There will also be custom streams, in order to address the use case of using this functionality with `StreamingGenerator`, as well as enabling all the other use cases.

I'll close this issue out with the 0.9 release.
